### DISCUSSION TITLE
How to deal with multiple versions of `time` when using macros

### DISCUSSION THREAD

**Discussion by paolobarbolini** [time-rs/time#356] [Question] [answered]
In https://github.com/launchbadge/sqlx/pull/1455 I'm trying to upgrade sqlx to time 0.3, but it was decided to keep support for time 0.2. Time 0.3 is imported as `time_03` and time 0.2 is imported as `time_02`.

Macros though seem to expect `time` not to have been renamed https://github.com/time-rs/time/blob/d700977ce9bc34013ee9841e122edbc807f410e4/time-macros/src/datetime.rs#L47

How can we fix this?

**Comment by jhpratt** [comment, marked as answer]
Yeah, this isn't currently possible. There is no equivalent of `$crate` for proc macros, and this would need the name for a _different_ crate.

**Reply by paolobarbolini** [reply]
What do you think if macros had a flag that simplify removed the `::` at the beginning of `time` imports? That way at the beginning of the file I could simply write `use time_03 as time;` and it would work. It would basically be an opt-in to the behavior currently present in all tokio macros:

> If you rename the Tokio crate in your dependencies this macro will not work. If you must rename the current version of Tokio because you’re also using an older version of Tokio, you must make the current version of Tokio available as `tokio` in the module where this macro is expanded.

https://docs.rs/tokio/1.12.0/tokio/attr.main.html#note

---

Another option could be to be able to have the option to specify under what name the macro can find `time`.

**Reply by jhpratt** [reply]
While that may work for something like tokio, "time" is something that could reasonably appear as a module name locally, which would naturally cause ambiguities. As to accepting something that could specify where `time` is found after renaming, yes, it's possible, but I've no idea what such syntax would be. It would necessarily have to be passed at _every_ call site, mind you.
